pkg/pbq/pbqmanager.go

```diff
@@ -0,0 +1,164 @@
+package pbq
+
+import (
+	"context"
+	"errors"
+	dfv1 "github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
+	"github.com/numaproj/numaflow/pkg/pbq/store"
+	"github.com/numaproj/numaflow/pkg/pbq/store/memory"
+	"github.com/numaproj/numaflow/pkg/pbq/store/noop"
+	"github.com/numaproj/numaflow/pkg/shared/logging"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sync"
+	"time"
+)
+
+var NonExistentPBQErr error = errors.New("missing PBQ for the partition")
+
+// Manager helps in managing the lifecycle of PBQ instances
+type Manager struct {
+	options *store.Options
+	pbqMap  map[string]*PBQ
+	log     *zap.SugaredLogger
+}
+
+// NewManager returns new instance of manager
+// We don't intend this to be called by multiple routines.
+func NewManager(ctx context.Context, opts ...store.SetOption) (*Manager, error) {
+	options := store.DefaultOptions()
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt(options); err != nil {
+				return nil, err
+			}
+		}
+	}
+	pbqManager := &Manager{
+		pbqMap:  make(map[string]*PBQ),
+		options: options,
+		log:     logging.FromContext(ctx),
+	}
+	return pbqManager, nil
+}
+
+// ListPartitions returns all the pbq instances
+func (m *Manager) ListPartitions() []*PBQ {
+
+	pbqList := make([]*PBQ, len(m.pbqMap))
+	i := 0
+	for _, val := range m.pbqMap {
```

Comment:
You need a lock to access the map.

Reply:
yes, I have added

Reply:
why do we need this lock? how many managers will be active at any point? is manager supposed to be accessed concurrently? 

One could argue that if `GetPQB` returns NonExistentPBQ, the entire PBQ could be managed through a go-routine, and hence a lock is required. 

Please document this thought in the `Manager` struct.

Reply:
I have documented this